packages/react-hooks/src/QuestionAnswerProvider.tsx

```diff
@@ -0,0 +1,104 @@
+import type { Agent } from '@aries-framework/core'
+import type {
+  QuestionAnswerRecord,
+  QuestionAnswerStateChangedEvent,
+  QuestionAnswerEventTypes,
+} from '@aries-framework/question-answer'
+import type { PropsWithChildren } from 'react'
+
+import { createContext, useState, useEffect, useContext, useMemo } from 'react'
+import * as React from 'react'
+
+interface QuestionAnswerContextInterface {
+  loading: true | 'no-qa' | 'yes-qa'
+  questionAnswerMessages: QuestionAnswerRecord[]
+}
+
+const QuestionAnswerContext = createContext<QuestionAnswerContextInterface | undefined>(undefined)
+
+const checkLoading = () => {
+  const questionAnswerContext = useContext(QuestionAnswerContext)
+  if (questionAnswerContext!.loading === 'no-qa') {
+    throw new Error('Question Answer hooks can only be used if Question Answer module is configured.')
+  }
+}
+
+export const useQuestionAnswer = (): { questionAnswerMessages: QuestionAnswerRecord[] } => {
+  const questionAnswerContext = useContext(QuestionAnswerContext)
+  if (!questionAnswerContext) {
+    throw new Error('useQuestionAnswer must be used within a QuestionAnswerContextProvider')
+  }
+  checkLoading()
+  return questionAnswerContext
+}
+
+export const useQuestionAnswerByConnectionId = (connectionId: string): QuestionAnswerRecord[] => {
+  const { questionAnswerMessages } = useQuestionAnswer()
+  const messages = useMemo(
+    () => questionAnswerMessages.filter((m: QuestionAnswerRecord) => m.connectionId === connectionId),
+    [questionAnswerMessages, connectionId]
+  )
+  return messages
+}
+
+export const useQuestionAnswerById = (id: string): QuestionAnswerRecord | undefined => {
+  const { questionAnswerMessages } = useQuestionAnswer()
+  return questionAnswerMessages.find((c: QuestionAnswerRecord) => c.id === id)
+}
+
+interface Props {
+  agent: Agent | undefined
+}
+
+const QuestionAnswerProvider: React.FC<PropsWithChildren<Props>> = ({ agent, children }) => {
+  const [questionAnswerState, setQuestionAnswerState] = useState<QuestionAnswerContextInterface>({
+    questionAnswerMessages: [],
+    loading: true,
+  })
+
+  const setInitialState = async () => {
+    if (agent) {
+      if (agent.modules.questionAnswer!) {
+        const questionAnswerMessages = await agent.modules.questionAnswer.getAll()
+        setQuestionAnswerState({ questionAnswerMessages, loading: 'yes-qa' })
+      } else {
+        setQuestionAnswerState({ questionAnswerMessages: [], loading: 'no-qa' })
```

Comment:
It feels a bit weird to me to have a value of `yes/no-qa` for the `loading` property.

I think we should only load the question answer provider if the question answer module is registered. For now, I think we can just add the quetsion answer module as a dependency and check if it's enabled? It means we do add the dependency, but it's a small pacakge, so it's fine I think.

That way we can do:
1. add an util method that checks whether the question/answer module is registered. We can do this as follows:
  - `Object.values(agent.dependencyManager.registeredModules).find((module) => module instanceof QuestionAnswerModule)`
2. In the `AgentProvider.tsx` we create a value that stores whether the module is enabled:
  - `const isQuestionAnswerModuleEnabled = useMemo(() => isQuestionAnswerModuleEnabled(agent), [agent])`
3. in the render of the `AgentProvider` we only add the question answer provider if the qa module is enabled.
4. If the qa provider is used, we can assume the module is always loaded. Also, we should not assume the module is loaded under `questionAnswer` as you can register the module under any key you wish. Rather, we can inject it:
  - `const questionAnswerApi = agent.dependencyManager.resolve(QuestionAnswerApi)`

Reply:
This is a good point, I'm going to check with James and get back to you.